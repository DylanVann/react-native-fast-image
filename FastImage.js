import React, { Component } from 'react'
import PropTypes from 'prop-types'
import {
  Image,
  NativeModules,
  requireNativeComponent,
} from 'react-native'
import ViewPropTypes from './utils/ViewPropTypes'

const resolveAssetSource = require('react-native/Libraries/Image/resolveAssetSource')

const FastImageViewNativeModule = NativeModules.FastImageView

class FastImage extends Component {
  setNativeProps(nativeProps) {
    this._root.setNativeProps(nativeProps)
  }

  render() {
    const {
      source,
      onLoadStart,
      onProgress,
      onLoad,
      onError,
      onLoadEnd,
      ...props
    } = this.props

    const resolvedSource = resolveAssetSource(source)
    return (
      <FastImageView
        ref={e => (this._root = e)}
        {...props}
        source={resolvedSource}
        onFastImageLoadStart={onLoadStart}
        onFastImageProgress={onProgress}
        onFastImageLoad={onLoad}
        onFastImageError={onError}
        onFastImageLoadEnd={onLoadEnd}
      />
    )
  }
}

FastImage.resizeMode = {
  contain: 'contain',
  cover: 'cover',
  stretch: 'stretch',
  center: 'center',
}

FastImage.priority = {
  low: 'low',
  normal: 'normal',
  high: 'high',
}

FastImage.preload = sources => {
  FastImageViewNativeModule.preload(sources)
}

FastImage.defaultProps = {
  resizeMode: FastImage.resizeMode.cover,
}

const FastImageSourcePropType = PropTypes.shape({
  uri: PropTypes.string,
  headers: PropTypes.objectOf(PropTypes.string),
  priority: PropTypes.oneOf(Object.keys(FastImage.priority)),
})

FastImage.propTypes = {
  ...ViewPropTypes,
  source: PropTypes.oneOfType([FastImageSourcePropType, PropTypes.number]),
  onLoadStart: PropTypes.func,
  onProgress: PropTypes.func,
  onLoad: PropTypes.func,
  onError: PropTypes.func,
  onLoadEnd: PropTypes.func,
}

const FastImageView = requireNativeComponent('FastImageView', FastImage, {
  nativeOnly: {
    onFastImageLoadStart: true,
    onFastImageProgress: true,
    onFastImageLoad: true,
    onFastImageError: true,
    onFastImageLoadEnd: true
  },
})

export default FastImage
