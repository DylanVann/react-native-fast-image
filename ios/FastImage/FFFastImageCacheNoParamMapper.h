//
//  FFFastImageCacheNoParamMapper.h
//  FastImage
//
//  Created by Kamil Badyla on 06/03/2019.
//  Copyright Â© 2019 vovkasm. All rights reserved.
//

#import <Foundation/Foundation.h>

NS_ASSUME_NONNULL_BEGIN

@interface FFFastImageCacheNoParamMapper : NSObject

+ (instancetype)shared;
- (void)add:(NSURL*)url;
- (void)remove:(NSURL*)url;

@end

NS_ASSUME_NONNULL_END
