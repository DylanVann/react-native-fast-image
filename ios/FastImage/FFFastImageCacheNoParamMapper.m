//
//  FFFastImageCacheNoParamMapper.m
//  FastImage
//
//  Created by Kamil Badyla on 06/03/2019.
//  Copyright Â© 2019 vovkasm. All rights reserved.
//

#import "FFFastImageCacheNoParamMapper.h"
#import <SDWebImage/SDWebImageManager.h>

@implementation NSURL (StaticUrl)

- (NSURL*)staticURL {
	return [[NSURL alloc] initWithScheme:self.scheme host:self.host path:self.path];
}

@end

@interface FFFastImageCacheNoParamMapper ()

@property (strong) NSMutableSet *staticUrls;

@end

@implementation FFFastImageCacheNoParamMapper

+ (instancetype)shared {
	static FFFastImageCacheNoParamMapper *_shared = nil;
	static dispatch_once_t onceToken;
	dispatch_once(&onceToken, ^{
		_shared = [FFFastImageCacheNoParamMapper new];
	});
	
	return _shared;
}

- (id)init {
	self = [super init];
	if (self) {
		_staticUrls = [NSMutableSet new];
		[[SDWebImageManager sharedManager] setCacheKeyFilter:^NSString * _Nullable(NSURL * _Nullable url) {
			NSString *staticURLString = [[url staticURL] absoluteString];
			if ([_staticUrls containsObject:staticURLString]) {
				return staticURLString;
			}
			return url.absoluteString;
		}];
	}
	return self;
}

- (void)add:(NSURL*)url {
	[_staticUrls addObject:[url staticURL].absoluteString];
}

- (void)remove:(NSURL*)url {
	[_staticUrls removeObject:[url staticURL].absoluteString];
}

@end
