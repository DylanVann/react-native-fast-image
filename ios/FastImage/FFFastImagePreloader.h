#import "FFFastImageSource.h"
#import <Foundation/Foundation.h>
#import <SDWebImage/SDWebImagePrefetcher.h>

@interface FFFastImagePreloader : SDWebImagePrefetcher

@property (nonatomic, readonly) NSNumber* id;

@end
