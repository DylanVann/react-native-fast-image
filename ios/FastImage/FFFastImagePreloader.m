#import "FFFastImagePreloader.h"
#import "FFFastImageSource.h"

static int instanceCounter = 0;

@implementation FFFastImagePreloader

-(instancetype) init {
    if (self = [super init]) {
        instanceCounter ++;
        _id = [NSNumber numberWithInt:instanceCounter];
    }
    return self;
}

@end
