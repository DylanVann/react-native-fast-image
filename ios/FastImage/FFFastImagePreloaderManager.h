#import <React/RCTBridgeModule.h>
#import <React/RCTEventEmitter.h>
#import <SDWebImage/SDWebImagePrefetcher.h>
#import <SDWebImage/SDWebImageDownloader.h>
#import <SDWebImage/SDWebImageManager.h>
#import <SDWebImage/SDImageCache.h>

@interface FFFastImagePreloaderManager : RCTEventEmitter <RCTBridgeModule, SDWebImagePrefetcherDelegate>

@end
