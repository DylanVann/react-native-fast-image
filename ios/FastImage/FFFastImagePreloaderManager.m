#import "FFFastImagePreloaderManager.h"
#import "FFFastImagePreloader.h"
#import "FFFastImageSource.h"

@implementation FFFastImagePreloaderManager
{
    bool _hasListeners;
    NSMutableDictionary* _preloaders;
}

RCT_EXPORT_MODULE(FastImagePreloaderManager);

- (dispatch_queue_t)methodQueue
{
    return dispatch_queue_create("com.dylanvann.fastimage.FastImagePreloaderManager", DISPATCH_QUEUE_SERIAL);
}

+ (BOOL)requiresMainQueueSetup
{
    return YES;
}

-(instancetype) init {
    if (self = [super init]) {
        _preloaders = [[NSMutableDictionary alloc] init];
    }
    return self;
}

- (NSArray<NSString *> *)supportedEvents
{
    return @[@"fffastimage-progress", @"fffastimage-complete"];
}

- (void) imagePrefetcher:(nonnull SDWebImagePrefetcher *)imagePrefetcher
 didFinishWithTotalCount:(NSUInteger)totalCount
            skippedCount:(NSUInteger)skippedCount
{
    NSNumber* id = ((FFFastImagePreloader*) imagePrefetcher).id;
    [_preloaders removeObjectForKey:id];
    [self sendEventWithName:@"fffastimage-complete"
                       body:@{ @"id": id, @"finished": [NSNumber numberWithLong:totalCount], @"skipped": [NSNumber numberWithLong:skippedCount]}
    ];
}

- (void) imagePrefetcher:(nonnull SDWebImagePrefetcher *)imagePrefetcher
          didPrefetchURL:(nullable NSURL *)imageURL
           finishedCount:(NSUInteger)finishedCount
              totalCount:(NSUInteger)totalCount
{
    NSNumber* id = ((FFFastImagePreloader*) imagePrefetcher).id;
    [self sendEventWithName:@"fffastimage-progress"
                       body:@{ @"id": id, @"finished": [NSNumber numberWithLong:finishedCount], @"total": [NSNumber numberWithLong:totalCount] }
    ];
}

RCT_EXPORT_METHOD(createPreloader:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    FFFastImagePreloader* preloader = [[FFFastImagePreloader alloc] init];
    preloader.delegate = self;
    _preloaders[preloader.id] = preloader;
    resolve(preloader.id);
}

RCT_EXPORT_METHOD(preload:(nonnull NSNumber*)preloaderId sources:(nonnull NSArray<FFFastImageSource *> *)sources) {
    NSMutableArray *urls = [NSMutableArray arrayWithCapacity:sources.count];
    
    [sources enumerateObjectsUsingBlock:^(FFFastImageSource * _Nonnull source, NSUInteger idx, BOOL * _Nonnull stop) {
        [source.headers enumerateKeysAndObjectsUsingBlock:^(NSString *key, NSString* header, BOOL *stop) {
            [[SDWebImageDownloader sharedDownloader] setValue:header forHTTPHeaderField:key];
        }];
        [urls setObject:source.url atIndexedSubscript:idx];
    }];
    
    FFFastImagePreloader* preloader = _preloaders[preloaderId];
    [preloader prefetchURLs:urls];
}

@end
