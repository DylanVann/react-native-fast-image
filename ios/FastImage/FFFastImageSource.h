#import <Foundation/Foundation.h>
#import <UIKit/UIKit.h>

typedef NS_ENUM(NSInteger, FFFPriority) {
    FFFPriorityLow,
    FFFPriorityNormal,
    FFFPriorityHigh
};

typedef NS_ENUM(NSInteger, FFFCacheControl) {
    FFFCacheControlImmutable,
    FFFCacheControlWeb,
    FFFCacheControlCacheOnly
};

// Object containing an image uri and metadata.
@interface FFFastImageSource : NSObject

// uri for image, or base64
@property (nonatomic) NSURL* url;
// priority for image request
@property (nonatomic) FFFPriority priority;
// headers for the image request
@property (nonatomic) NSDictionary *headers;
// cache control mode
@property (nonatomic) FFFCacheControl cacheControl;

- (instancetype)initWithURL:(NSURL *)url
                   priority:(FFFPriority)priority
                    headers:(NSDictionary *)headers
               cacheControl:(FFFCacheControl)cacheControl;

@end
