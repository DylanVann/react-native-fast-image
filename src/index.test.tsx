import { StyleSheet, Platform, NativeModules } from 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'
import FastImage from './index'

const style = StyleSheet.create({ image: { width: 44, height: 44 } })

describe('FastImage (iOS)', () => {
    beforeAll(() => {
        Platform.OS = 'ios'
        NativeModules.FastImageView = {
            preload: Function.prototype,
            clearMemoryCache: Function.prototype,
            clearDiskCache: Function.prototype,
        }
    })

    it('renders', () => {
        const tree = renderer
            .create(
                <FastImage
                    source={{
                        uri: 'https://facebook.github.io/react/img/logo_og.png',
                        headers: {
                            token: 'someToken',
                        },
                        priority: FastImage.priority.high,
                    }}
                    style={style.image}
                />,
            )
            .toJSON()

        expect(tree).toMatchSnapshot()
    })

    it('renders a normal Image when not passed a uri', () => {
        const tree = renderer
            .create(
                <FastImage
                    source={require('../ReactNativeFastImageExampleServer/pictures/jellyfish.gif')}
                    style={style.image}
                />,
            )
            .toJSON()

        expect(tree).toMatchSnapshot()
    })

    it('renders Image with fallback prop', () => {
        const tree = renderer
            .create(
                <FastImage
                    source={require('../ReactNativeFastImageExampleServer/pictures/jellyfish.gif')}
                    style={style.image}
                    fallback
                />,
            )
            .toJSON()

        expect(tree).toMatchSnapshot()
    })

    it('renders defaultSource', () => {
        const tree = renderer
            .create(
                <FastImage
                    defaultSource={require('../ReactNativeFastImageExampleServer/pictures/jellyfish.gif')}
                    style={style.image}
                />,
            )
            .toJSON()

        expect(tree).toMatchSnapshot()
    })

    it('runs static functions', () => {
        FastImage.preload([
            {
                uri: 'https://facebook.github.io/react/img/logo_og.png',
                headers: {
                    token: 'someToken',
                },
                priority: FastImage.priority.high,
            },
        ])
        FastImage.clearMemoryCache()
        FastImage.clearDiskCache()
    })
})

describe('FastImage (Android)', () => {
    beforeAll(() => {
        Platform.OS = 'android'
    })

    it('renders a normal defaultSource', () => {
        const tree = renderer
            .create(
                <FastImage
                    defaultSource={require('../ReactNativeFastImageExampleServer/pictures/jellyfish.gif')}
                    style={style.image}
                />,
            )
            .toJSON()

        expect(tree).toMatchSnapshot()
    })

    it('renders a normal defaultSource when fails to load source', () => {
        const tree = renderer
            .create(
                <FastImage
                    defaultSource={require('../ReactNativeFastImageExampleServer/pictures/jellyfish.gif')}
                    source={{
                        uri: 'https://www.google.com/image_does_not_exist.png',
                    }}
                    style={style.image}
                />,
            )
            .toJSON()

        expect(tree).toMatchSnapshot()
    })

    it('renders a non-existing defaultSource', () => {
        const tree = renderer
            .create(<FastImage defaultSource={12345} style={style.image} />)
            .toJSON()

        expect(tree).toMatchSnapshot()
    })
})
