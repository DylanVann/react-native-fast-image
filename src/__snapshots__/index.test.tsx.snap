// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FastImage renders correctly. 1`] = `
<View
  style={
    Array [
      Object {
        "overflow": "hidden",
      },
      Object {
        "height": 44,
        "width": 44,
      },
    ]
  }
>
  <FastImageView
    resizeMode="cover"
    source={
      Object {
        "headers": Object {
          "token": "someToken",
        },
        "priority": "high",
        "uri": "https://facebook.github.io/react/img/logo_og.png",
      }
    }
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
</View>
`;

exports[`Renders Image with fallback prop. 1`] = `
<View
  style={
    Array [
      Object {
        "overflow": "hidden",
      },
      Object {
        "height": 44,
        "width": 44,
      },
    ]
  }
>
  <Image
    resizeMode="cover"
    source={
      Object {
        "testUri": "../../../ReactNativeFastImageExampleServer/pictures/jellyfish.gif",
      }
    }
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
</View>
`;

exports[`Renders a normal Image when not passed a uri. 1`] = `
<View
  style={
    Array [
      Object {
        "overflow": "hidden",
      },
      Object {
        "height": 44,
        "width": 44,
      },
    ]
  }
>
  <FastImageView
    resizeMode="cover"
    source={
      Object {
        "testUri": "../../../ReactNativeFastImageExampleServer/pictures/jellyfish.gif",
      }
    }
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
</View>
`;
